Source/Controllers/Other/SecureBookmarkManager.swift

```diff
@@ -0,0 +1,308 @@
+//
+//  SecureBookmarkManager.swift
+//  Sequel Ace
+//
+//  Created by James on 7/12/2020.
+//  Copyright © 2020 Sequel-Ace. All rights reserved.
+//
+
+import Firebase
+import Foundation
+import os.log
+
+/*
+ reRequestSecureAccess
+ addBookmark
+ handle bookmarkDataIsStale
+ revokeBookmark
+ stopAllSecurityScopedAccess
+
+ */
+
+@objc final class SecureBookmarkManager: NSObject {
+    @objc static let sharedInstance = SecureBookmarkManager()
+
+    @objc var bookmarks: [Dictionary<String, Data>] = []
+    @objc var staleBookmarks: [String] = []
+
+    private var resolvedBookmarks: [URL] = []
+    private let URLBookmarkResolutionWithSecurityScope = URL.BookmarkResolutionOptions(rawValue: 1 << 10)
+    private let log: OSLog
+    private let prefs: UserDefaults = UserDefaults.standard
+    private var observer: NSKeyValueObservation?
+
+    private var iChangedTheBookmarks: Bool = false
+
+    override init() {
+        log = OSLog(subsystem: Bundle.main.bundleIdentifier!, category: "secureBookmarks")
+
+        os_log("SecureBookmarkManager init.", log: log, type: .info)
+        Crashlytics.crashlytics().log("SecureBookmarkManager init.")
+
+        super.init()
+
+        // this manager *should* be the only thing changing the bookmarks pref, but in case...
+        observer = UserDefaults.standard.observe(\.SPSecureBookmarks, options: [.new, .old], changeHandler: { _, change in
+
+            if self.iChangedTheBookmarks == false {
+                os_log("SPSecureBookmarks changed NOT by SecureBookmarkManager.", log: self.log, type: .debug)
+                Crashlytics.crashlytics().log("SPSecureBookmarks changed NOT by SecureBookmarkManager.")
+                self.bookmarks.removeAll()
+                self.bookmarks = change.newValue!
+            }
+            // reset
+            self.iChangedTheBookmarks = false
+
+            // post notificay for ConnectionController
+            NotificationCenter.default.post(name: Notification.Name(NSNotification.Name.SPBookmarksChanged.rawValue), object: self)
+
+        })
+
+        // FIXME: @Kaspik need help ... need a guard or if let, don't want the default value...
```

Comment:
Just flagging the fixme

Reply:
It's already fixed, just didn't delete the docu

Reply:
sorted.